pkg/apis/scaler/v1alpha2/types_demand.go

```diff
@@ -0,0 +1,100 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha2
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Demand represents currently unschedulable resources.
+type Demand struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   DemandSpec   `json:"spec,omitempty"`
+	Status DemandStatus `json:"status,omitempty"`
+}
+
+// DemandSpec represents the units and status of a Demand resource
+type DemandSpec struct {
+	Units         []DemandUnit `json:"units"`
+	InstanceGroup string       `json:"instance-group"`
+
+	// IsLongLived changes the lifecycle for a demand from
+	// ephemeral and immutable to long-lived and mutable.
+	// This is useful to set a buffer in an instance-group:
+	// an amount of compute resources that is left unused
+	// but ready for quick reservation should there be need.
+	IsLongLived bool `json:"is-long-lived"`
+}
+
+// DemandStatus represents the status a demand object is in
+type DemandStatus struct {
+	// Phase denotes the demand phase.
+	Phase string `json:"phase"`
+	// LastTransitionTime denotes the last transition time of the demand phase.
+	// If left empty, defaults to the creation time of the demand.
+	// +optional
+	LastTransitionTime metav1.Time `json:"last-transition-time,omitempty"`
+}
+
+// ResourceList is a set of (resource name, quantity) pairs.
+type ResourceList map[corev1.ResourceName]resource.Quantity
+
+// DemandUnit represents a single unit of demand as a count of resources requirements
+type DemandUnit struct {
+	Resources ResourceList `json:"resources"`
+	Count     int          `json:"count"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DemandList is a list of Demand resources
+type DemandList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []Demand `json:"items"`
+}
+
+// CPU Returns the CPU demand if specified
+func (r *ResourceList) CPU() *resource.Quantity {
+	if val, ok := (*r)[ResourceCPU]; ok {
+		return &val
+	}
+	return &resource.Quantity{Format: resource.DecimalSI}
+}
+
+// Memory Returns the Memory demand if specified
+func (r *ResourceList) Memory() *resource.Quantity {
+	if val, ok := (*r)[ResourceMemory]; ok {
+		return &val
+	}
+	return &resource.Quantity{Format: resource.DecimalSI}
+}
+
+// NvidiaGPU Returns the GPU demand if specified
+func (r *ResourceList) NvidiaGPU() *resource.Quantity {
```

Comment:
I don't think we should be adding anything specific to GPUs here.

Reply:
We should declare the resources that spark-scheduler supports. If this + resourceName was not declared users wouldn't know nvidia gpus are supported